azurelinuxagent/common/protocol/wire.py

```diff
@@ -802,29 +799,45 @@ def update_goal_state(self, forced=False):
 
     def update_extensions_goal_state(self):
         try:
-            correlation_id = str(uuid.uuid4())
-            url, headers = self.get_host_plugin().get_vm_settings_request(correlation_id)
-            etag = self.get_etag()
-            if etag is not None:
-                headers['if-none-match'] = etag
-
-            vm_settings, response_headers = self.fetch(url, headers, ok_codes=[httpclient.OK, httpclient.NOT_MODIFIED])
-
-            # The response includes an etag if and only if the VM settings change.
-            response_etag = None
-            for h in response_headers:
-                if h[0].lower() == 'etag':
-                    response_etag = h[1]
+            for retry in range(3):
+                correlation_id = str(uuid.uuid4())
+                url, headers = self.get_host_plugin().get_vm_settings_request(correlation_id)
+                etag = self.get_etag()
+                if etag is not None:
+                    headers['if-none-match'] = etag
+
+                # TODO: Remove this log statement when invoking this method on each iteration of the goal state loop (currently it is invoked only on a new goal state)
+                logger.info("Fetching extensions goal state [correlation ID: {0} eTag: {1}]", correlation_id, etag)
+
+                vm_settings, response_headers = self.fetch(url, headers, ok_codes=[httpclient.OK, httpclient.NOT_MODIFIED])
+
+                # on error, fetch() logs the original exception and returns None as content
+                if vm_settings is None:
+                    raise Exception("Failed to retrieve the extensions goal state (vmSettings), the error was reported previously in the log.")
+
+                # The response includes an etag if and only if the VM settings change.
+                response_etag = None
+                for h in response_headers:
+                    if h[0].lower() == 'etag':
+                        response_etag = h[1]
+                        break
+
+                # TODO: Remove this log statement when invoking this method on each iteration of the goal state loop (currently it is invoked only on a new goal state)
+                logger.info("Fetched extensions goal state [correlation ID: {0} eTag: {1}]", correlation_id, response_etag)
+
+                if response_etag is None:
+                    logger.info("Expected a new goal state; will retry after a short delay...")
+                    time.sleep(5)
+                else:
+                    logger.info("Fetched new extensions goal state [correlation ID: {0} eTag: {1}]", correlation_id, response_etag)
```

Comment:
What's the reasoning behind having both this logging statement and the one above? they look like they're providing the same information.

Reply:
the statement  at line 826 will be removed as part of the TODO. This one will stay.